ESQL: Add parsing for a LOOKUP command

nik9000 opened the pull request:
This command will serve as a sort of "inline" enrich. This commit itself is mostly antlr generated code and paranoid tests that the new `LOOKUP` keyword doesn't clash with any variables named `lookup`.

I've also marked our ANTLR generated files as `linguist-generated` which causes them to be hidden by default in github's UI. You can still click a button to see them if you like. See https://docs.github.com/en/repositories/working-with-files/managing-files/customizing-how-changed-files-appear-on-github

nik9000 committed:
ESQL: Add parsing for a LOOKUP command

This command will serve as a sort of "inline" enrich. This commit itself
is mostly antlr generated code and paranoid tests that the new `LOOKUP`
keyword doesn't clash with any variables named `lookup`.

elasticsearchmachine commented:
Pinging @elastic/es-analytical-engine (Team:Analytics)

costin reviewed (approved):
LGTM

nik9000 commented:
BWC failure is real - LOOKUP can't be used there. Easy enough.

nik9000 committed:
Better test

alex-spies commented on x-pack/plugin/esql/src/main/antlr/EsqlBaseParser.g4:
```diff
@@ -268,3 +273,7 @@ enrichCommand
     ;
+
+lookupCommand
+    : LOOKUP tableName=qualifiedNamePattern ON matchFields=qualifiedNamePatterns
```
I'm really not sure if this should use a qualified name pattern for the table name. IMHO the table name should be somewhat in line with `FROM`'s index name and `ENRICH`'s policy name. Qualified name patterns are usually used for field names, so we can do stuff like
```
KEEP `some-thing`.`something-else`.yet_something_else
```

`FROM` uses an `INDEX_UNQUOTED_IDENTIFIER`, `ENRICH` a similar `ENRICH_POLICY_NAME_BODY`.

Maybe we just align with `FROM` here and use `INDEX_UNQUOTED_IDENTIFIER` as well, unless table names need to be more general than that?

nik9000 replied:
I'll give it a go.

alex-spies reviewed (commented):
Looks good; I think we could make the table name less general, though. (See below.)

nik9000 committed:
Merge branch 'main' into lookup_syntax

nik9000 committed:
Smaller

nik9000 committed:
ESQL: Remove `nested` shims

This removes the shim layers for `nested` that we inherited from QL.
ESQL doesn't support nested yet and when it does support nested its not
going to use the same tricks that QL uses.

nik9000 committed:
Revert "ESQL: Remove `nested` shims"

This reverts commit e50d09dbca6da36c2238fe2de7b4345591776363.

nik9000 committed:
This time is right

ChrisHegarty commented on .gitattributes:
```diff
@@ -3,3 +3,8 @@ CHANGELOG.asciidoc  merge=union
 # Windows
 build-tools-internal/src/test/resources/org/elasticsearch/gradle/internal/release/*.asciidoc text eol=lf
+
+x-pack/plugin/esql/src/main/antlr/*.tokens linguist-generated=true
+x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/parser/*.interp linguist-generated=true
+x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/parser/EsqlBaseLexer*.java linguist-generated=true
+x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/parser/EsqlBaseParser*.java linguist-generated=true
```
ha! interesting. what does the IDE do, if anything, with this ?

nik9000 replied:
So far as I can tell, nothing.
